rust/lance-core/src/utils/tokio.rs

```diff
@@ -1,16 +1,35 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-FileCopyrightText: Copyright The Lance Authors
 
+use std::time::Duration;
+
 use crate::Result;
 
 use futures::{Future, FutureExt};
 use tokio::runtime::{Builder, Runtime};
 use tracing::Span;
 
+fn get_num_compute_intensive_cpus() -> usize {
+    if num_cpus::get() <= *IO_CORE_RESERVATION {
+        log::warn!(
+            "Number of CPUs is less than or equal to the number of IO core reservations. \
+            This is not a supported configuration. using 1 CPU for compute intensive tasks."
+        );
+        return 1;
+    }
+
+    num_cpus::get() - *IO_CORE_RESERVATION
+}
+
 lazy_static::lazy_static! {
+    pub static ref IO_CORE_RESERVATION: usize = std::env::var("LANCE_IO_CORE_RESERVATION").unwrap_or("2".to_string()).parse().unwrap();
+
     pub static ref CPU_RUNTIME: Runtime = Builder::new_multi_thread()
         .thread_name("lance-cpu")
-        .max_blocking_threads(num_cpus::get())
+        .max_blocking_threads(get_num_compute_intensive_cpus())
+        .worker_threads(1)
```

Comment:
Are we also restricting I/O to a single thread?

Reply:
we never schedule IO on this runtime, so this is fine.